### DISCUSSION TITLE
Duplicates showing up in playlist

### DISCUSSION THREAD

**Discussion by hahahahywel** [Chimildic/goofy#244] [Помощь с ошибкой] [answered]
Hi there,

I have a playlist that is sorted based on when I last played a track with the most recently played going to the bottom of the playlist. This has worked fine for a while but there are two songs which are causing me a headache, both conveniently by the same artist: ALASKALASKA.

```
function LessPlayedAdded() {

// get songs from play history
 let play_hist = Source.getTracks([
  { name: 'Play History', id: '6J3OdPDClqYrXg7SGlPr1D' },
  { name: 'Play History archive 3', id: '4e09pxzOwYSYtSrvLqYv0A' },
  { name: 'Play History archive 2', id: '6Mtlf3DLANK1sxcEtlO1XQ' },
]);

// get 10000 played songs
 Selector.keepFirst(play_hist,10000);

// get liked songs
 let liked_songs = Source.getSavedTracks(50);

// get unavailable tracks
 let unavailable = Source.getPlaylistTracks('Unavailable', '5poaVEvOZdAEeH3oQAXe1v');



// remove duplicates
 Filter.dedupTracks(play_hist)

// reverse play history
 Order.reverse(play_hist);

// remove non-favourite songs from ordered play history
 Filter.removeTracks(play_hist, liked_songs, true);
 Filter.removeTracks(play_hist, unavailable);




 Playlist.saveWithReplace({
     id: '4xIRcWOCBmNU0Z1YygrarZ',
     name: 'Less Played Recently Added',
     tracks: play_hist,
     randomCover: 'update',
   });

}
```

In the resulting array, there are songs called "Monster" and "Meateater" which appear twice. One copy of the songs follows the playlist's rules but the other just sits at the top of the playlist. Both copies have different URIs even though I have only liked one copy of each song. 

Monster: spotify:track:5ZpPxBttRqSRMa8aim3gzJ, spotify:track:4wD7e5tXLowG9XnANZIKo1
Meateater: spotify:track:3FBelUuzZ5IVjJVVbIwOIS, spotify:track:4QED6EwRbgMLDdbY2N3AVW
(The first URI for each song is the one that follows the rules, the other is the one that seems to be in error)

Any help would be appreciated because it's messing up my morning playlist and annoying me!

Thanks

**Comment by Chimildic** [comment, marked as answer]
When tracks are different _ids_ and the same _names_, [dedupTracks function](https://chimildic.github.io/goofy/#/reference/filter?id=deduptracks) looks at duration. If the offset is too much, the tracks are considered different. For example, `monster` has 3:56 and 3:53, `meateater` has 4:25 and 4:22. 

In order to remove such cases, the function has the second argument `offsetDurationMs` (by default 2000 ms). As you can see, the tracks have more than 2 seconds offset. Thus they hold on.

Just add some offset:
```
Filter.dedupTracks(play_hist, 5000)
```

**Reply by hahahahywel** [reply]
Thanks so much, the problem is now sorted! Can't believe I didn't see that before.

Thanks again
